### PR TITLE
#1262 deprecate the SpinSail

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #1262  

Briefly describe the changes proposed in this PR:

* Deprecated the SpinSail and added comments that explain why and recommend alternative solutions
* 
*

### CODE REVIEW

File: core/sail/sail-spin/src/main/java/org/eclipse/rdf4j/sail/spin/SpinSail.java [hunk 1]
```diff
@@ -29,6 +29,23 @@
...
+ * Deprecating the SpinSail has been discussed at https://github.com/eclipse/rdf4j/issues/1262 and can be summarized
+ * with there being no developers actively supporting it and that SPIN in itself is no longer recommended by
+ * TopQuadrant. Do not expect the SpinSail to scale in any way and expect simple delete operations to take seconds to
+ * complete. There are still a number of open issues in GitHub connected to the SpinSail, none of these are likely to
+ * get fixed.
+ * </p>
+ *
+ * @since 3.1.0 2019
+ */
+@Deprecated
```
Comment (hmottestad) [inline core/sail/sail-spin/src/main/java/org/eclipse/rdf4j/sail/spin/SpinSail.java]: Please read the above copy and let me know what you think.

File: core/sail/sail-spin/src/main/java/org/eclipse/rdf4j/sail/spin/SpinSail.java [hunk 2]
```diff
@@ -29,6 +29,23 @@
 import java.util.ArrayList;
 import java.util.List;
 
+/**
+ * <p>
+ * The SpinSail is currently deprecated. If you are using SPIN to validate your data then it is recommended to move to
+ * SHACL with the ShaclSail. SHACL does not support any of the SPARQL inference features of SPIN, however some were
+ * proposed in "SHACL Advanced Features" but were not made a part of the W3C Recommendation in 2017.
```
Comment (jeenbroekstra) [inline core/sail/sail-spin/src/main/java/org/eclipse/rdf4j/sail/spin/SpinSail.java]: I'd rephrase this:

 "SHACL does not support any of the SPARQL inference features of SPIN, however some were proposed in "SHACL Advanced Features" but were not made a part of the W3C Recommendation in 2017." 

to something shorter, like: "Currently, the <a href="https://www.w3.org/TR/shacl/">SHACL W3C Recommendation</a> only supports validation, and has no equivalent to SPIN's inference features."

**Comment by hmottestad** [comment]
Also need to change this: https://rdf4j.org/documentation/programming/spin/ to reflect that SPIN is no longer recommended.

File: core/sail/sail-spin/src/test/java/org/eclipse/rdf4j/sail/spin/SpinSailTest.java [hunk 3]
```diff
@@ -141,6 +147,39 @@ public void testSpinxRule() throws Exception {
...
+		conn.commit();
+		conn.begin();
+
+		conn.add(ex2, FOAF.KNOWS, ex3);
+		conn.add(ex3, FOAF.KNOWS, ex4);
+		conn.commit();
+
+		assertStatements("testDeepConstructRule-expected.ttl");
+
+	}
```
Comment (hmottestad) [inline core/sail/sail-spin/src/test/java/org/eclipse/rdf4j/sail/spin/SpinSailTest.java]: @jeenbroekstra 
I've added this test here to show the lack of transactional support in the SpinSail. Is there some way for us to make this kind of scenario function? Eg. by having the SpinSail always recompute every rule on every transaction?

Comment (hmottestad) [reply core/sail/sail-spin/src/test/java/org/eclipse/rdf4j/sail/spin/SpinSailTest.java]: Essentially, the current implementation only triggers a computation when a matching type statement is added. So if it is added in the first transaction, and later we add more data that should actually trigger the rule....then nothing is inferred.

Comment (hmottestad) [reply core/sail/sail-spin/src/test/java/org/eclipse/rdf4j/sail/spin/SpinSailTest.java]: It's actually a little bit better, it's triggered when a resource is modified if that resource has a matching type.
